torchdata/dataloader2/reading_service.py

```diff
@@ -235,3 +240,84 @@ def finalize(self) -> None:
         if self.persistent_workers and self.dl_ is not None and self.dl_._iterator is not None:
             self.dl_._iterator._shutdown_workers()  # type: ignore[attr-defined]
             self.dl_._iterator = None
+
+
+SHARED_SEED = "_dl_shared_seed"
+SHARED_SEED_COUNTER = "_dl_shared_seed_recv_cnt"
+
+
+class DistributedReadingService(ReadingServiceInterface):
+    def __init__(self, timeout: int = default_timeout_in_s, check_interval: float = default_check_interval):
+        if not (dist.is_available() and dist.is_initialized()):
+            raise RuntimeError("Torch Distributed is required to be initialized")
+        self._world_size: int = dist.get_world_size()
+        self._rank: int = dist.get_rank()
+        self._seed_generator = torch.Generator()
+        self._datapipe: Optional[DataPipe] = None
+        self._timeout: int = timeout
+        self._check_interval: float = check_interval
+
+    def initialize(self, datapipe: DataPipe) -> DataPipe:
+        torch.utils.data.graph_settings.apply_sharding(
+            datapipe,
+            self._world_size,
+            self._rank,
+        )
+        datapipe = datapipe.fullsync(self._timeout)
+        self._datapipe = datapipe
+        return datapipe
+
+    def initialize_iteration(self) -> None:
+        seed = self._share_seed()
+        self._seed_generator.manual_seed(seed)
+        assert self._datapipe is not None
+        self._datapipe = torch.utils.data.graph_settings.apply_shuffle_seed(
+            self._datapipe,
+            self._seed_generator,
+        )
```

Comment:
This should be changed after `initialize_iteration` is modified.

Reply:
I may be out of the loop but what is the planned change here?

Reply:
`initialize_iteration` should accept the RNG from `DataLoader2` to set randomness over the pipeline within RS.